src/gc/gc.cpp

```diff
@@ -11550,32 +11550,60 @@ void gc_heap::adjust_limit_clr (uint8_t* start, size_t limit_size,
         assert (heap_segment_used (seg) >= old_allocated);
     }
 #endif //BACKGROUND_GC
-    if ((seg == 0) ||
-        (start - plug_skew + limit_size) <= heap_segment_used (seg))
+
+    // we are going to clear a right-edge exclusive span [clear_start, clear_limit)  
+    // but will adjust for cases when object is ok to stay dirty or the space has not seen any use yet
+    // NB: the size and limit_size include syncblock, which is to the -1 of the object start 
+    //     that effectively shifts the allocation by `plug_skew`
+    uint8_t* clear_start = start - plug_skew;
+    uint8_t* clear_limit = start + limit_size - plug_skew;
+
+    if (flags & GC_ALLOC_ZEROING_OPTIONAL)
+    {
+        uint8_t* obj_start = acontext->alloc_ptr;
+        assert(start >= obj_start);
+        uint8_t* obj_end = obj_start + size - plug_skew;
+        assert(obj_end > clear_start);
+
+        // if clearing at the object start, clear the syncblock.
+        if(obj_start == start)
+        {
+            *(PTR_PTR)clear_start = 0;
```

Comment:
this is correct except in the case of GCHeap::AllocAlign8Common (only used by arm32) which might ask for one min obj more in size and allocate the object at alloc_ptr + min_obj. so you'd need to clear the sync block for that too.

Reply:
Hmm. I would expect failures on ARM32 because of this. Something saves us here.

---
In reply to: [278281628](https://github.com/dotnet/coreclr/pull/24096#discussion_r278281628) [](ancestors = 278281628)

Reply:
it's not like you are guaranteed to get failures - the memory for the sync block could very well be already 0; or the usage in the test just doesn't need to care about sync blocks.

Reply:
Right. There are many preconditions that need to happen to observe this as a failure. Far from any object will need re-alignment. Even fewer will get dirty syncblock and yet fewer will care.

---
In reply to: [278293831](https://github.com/dotnet/coreclr/pull/24096#discussion_r278293831) [](ancestors = 278293831)